### ISSUE TITLE
bcftools concat could use some improvements on header compatibility check

### ISSUE THREAD

**Issue by SHuang-Broad** [samtools/bcftools#1591]
The original error message
```bash
$ bcftools sort --output-type z --temp-dir local_tmp -o sample_name.clair.g.vcf.gz naive_concatedated_raw.vcf.gz
Writing to local_tmpGUdCXo
[W::vcf_parse] Contig 'chr17' is not defined in the header. (Quick workaround: index the file with tabix.)
Error encountered while parsing the input at chr17:1
Cleaning
```

Header of the unsorted VCF

```bash
$ grep "^#" <(zcat concatedated_raw.vcf.gz)
##fileformat=VCFv4.2##FILTER=<ID=PASS,Description="All filters passed">
##FILTER=<ID=LowQual,Description="Low quality variant">
##FILTER=<ID=RefCall,Description="Reference call">##INFO=<ID=P,Number=0,Type=Flag,Description="Result from pileup calling">
##INFO=<ID=F,Number=0,Type=Flag,Description="Result from full-alignment calling">
##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">##INFO=<ID=END,Number=1,Type=Integer,Description="End position (for use with symbolic alleles)">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read Depth">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth for each allele">
##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block">##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Phred-scaled genotype likelihoods rounded to the closest integer">
##FORMAT=<ID=AF,Number=1,Type=Float,Description="Estimated allele frequency in the range of [0,1]">
##contig=<ID=chr1,length=248956422>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  HG00514
```

And the relevant lines from the body

```bash
$ zcat concatedated_raw.vcf.gz | grep -m 2 -C 2 "^chr17"
chr1    248946422       .       G       <NON_REF>       0       .       END=248946422   GT:GQ:MIN_DP:PL 0/0:50:17:0,51,509
chr1    248946423       .       N       <NON_REF>       0       .       END=248956422   GT:GQ:MIN_DP:PL ./.:1:0:0,0,0
chr17   1       .       N       <NON_REF>       0       .       END=60000       GT:GQ:MIN_DP:PL ./.:1:0:0,0,0
chr17   60001   .       G       <NON_REF>       0       .       END=62982       GT:GQ:MIN_DP:PL 0/0:1:0:0,0,0
chr17   62983   .       G       <NON_REF>       0       .       END=62990       GT:GQ:MIN_DP:PL 0/0:4:1:0,3,29
chr17   62991   .       T       <NON_REF>       0       .       END=62991       GT:GQ:MIN_DP:PL ./.:0:1:29,3,0
```

It just turns out that the VCF to be sorted was generated via 

```bash
$ bcftools concat --naive -f all_raw_vcfs.txt --output-type v -o concatedated_raw.vcf.gz

Checking the headers of 21 files.
Done, the headers are compatible.
```
The VCFs listed in `all_raw_vcfs.txt` has partial listing of contigs, i.e. the contigs the caller sees in the input bam, which in turn was generated from my parallelization scheme that chops up a WGS bam into 21 size-balanced intervals.

And from the documentation of `bcftools concat`

> **-n, --naive**
    Concatenate VCF or BCF files without recompression. This is very fast but requires that all files are of the same type (all VCF or all BCF) and have the same headers. This is because all tags and chromosome names in the BCF body rely on the order of the contig and tag definitions in the header. A header check compatibility is performed and the program throws an error if it is not safe to use the option.

In summary, `bcftools concat` was checking the compatibility of the headers of VCFs, and missed the error that the headers, arguably, are in-compatible.

Thanks,
Steve
